sapicore/engine/network/__init__.py

```diff
@@ -0,0 +1,400 @@
+                self.graph.add_node(ensemble.identifier, reference=ensemble)
+
+            elif isinstance(ensemble, str):
+                if os.path.exists(ensemble):
+                    # add the ensemble reference to the graph as a node.
+                    ensemble, comp_cfg = self._object_from_configuration(path=ensemble, comp_type="ensemble")
+                    self.graph.add_node(ensemble.identifier, reference=ensemble)
+                else:
+                    logging.info(f"Could not add ensemble, invalid path given: {ensemble}")
+                    continue
+            else:
+                raise TypeError("Ensembles must be given as object references, dictionaries, or paths to YAML.")
+
+            # heterogenize modifies the object iff a `sweep` specification was provided in the model configuration.
+            ensemble.heterogenize(num_combinations=ensemble.num_units)
+
+        # update root node list.
+        self._find_roots()
+
+    def add_synapses(self, *args: Synapse | dict | str):
+        """Adds synapse edges to the network graph from paths to YAML, from pre-initialized
+        :class:`engine.synapse.Synapse` objects or children thereof (e.g.,`STDPSynapse`), or from
+        configuration dictionaries. User may mix and match if necessary.
+
+        Raises
+        ------
+        TypeError
+            If user attempts to pass an argument that is not a string path, a dictionary, or an instance derived
+            from :class:`synapse.Synapse`.
+
+        """
+        for synapse in args:
+            if isinstance(synapse, Synapse):
+                # handle the object arguments.
+                self.graph.add_edge(synapse.src_ensemble.identifier, synapse.dst_ensemble.identifier, reference=synapse)
+
+            elif type(synapse) is dict:
+                synapse, comp_cfg = self._object_from_configuration(cfg=synapse, comp_type="synapse")
+                self.graph.add_edge(synapse.src_ensemble.identifier, synapse.dst_ensemble.identifier, reference=synapse)
+
+            elif type(synapse) is str:
+                if os.path.exists(synapse):
+                    # add the synapse reference to the graph as a node.
+                    synapse, comp_cfg = self._object_from_configuration(path=synapse, comp_type="synapse")
+                    self.graph.add_edge(*synapse.identifier.split("->"), reference=synapse)
+                else:
+                    logging.info(f"Could not add synapse, invalid path given: {synapse}")
+                    continue
+            else:
+                raise TypeError("Synapses must be given as object references, dictionaries, or paths to YAML.")
+
+            # # heterogenize modifies the object iff a `sweep` specification was provided in the model configuration.
+            synapse.heterogenize(num_combinations=synapse.num_units)
+
+        # update root node list.
+        self._find_roots()
```

Comment:
The only snag is that build() isn't the only way to build a network. A user might have a few separate add_ensembles calls in their code, each having the potential to change node in-degrees (and therefore the default flow). They're also meant to be called with multiple *args. In practice (not in a troll scenario where the user is determined to choke their CPU), it's not an expensive operation--considering it's applied to layers and synapse matrices (it makes sense to have massive duplication within those objects).

Reply:
About build being not the only way, `reconfigure=True` should handle that. Because by default then it would reconfig on every call. Only build would make it so that the optimization is applied and it's False.

But, if you think it's not costly, that's fine.

Reply:
We can always do optimization later and trace execution time of all those function to see where we need to optimize things :)

Reply:
> by default then it would reconfig on every call. Only build would make it so that the optimization is applied and it's False.

If that's what you meant,  it makes sense. A very minor optimization but an optimization nonetheless.

Remember that instantiating and adding a bunch of singleton ensembles and synapses (the only scenario where this optimization would be felt) is not viable. You'd almost always have a single-digit or low double-digit number of ensemble and synapse objects, with massive duplication relegated to internal tensors.